R/server.R

```diff
@@ -561,14 +561,60 @@ runApp <- function(appDir=getwd(),
     .globals$options <- oldOptionSet
   },add = TRUE)
 
-  if (is.null(host) || is.na(host))
-    host <- '0.0.0.0'
-
   # Make warnings print immediately
   # Set pool.scheduler to support pool package
   ops <- options(warn = 1, pool.scheduler = scheduleTask)
   on.exit(options(ops), add = TRUE)
 
+  appParts <- as.shiny.appobj(appDir)
+
+  # The lines below set some of the app's running options, which
+  # can be:
+  #   - left unspeficied (in which case the arguments' default
+  #     values from `runApp` kick in);
+  #   - passed through `shinyApp`
+  #   - passed through `runApp` (this function)
+  #   - passed through both `shinyApp` and `runApp` (the latter
+  #     takes precedence)
+  #
+  # Matrix of possibilities:
+  # | IN shinyApp | IN runApp | result       | check                                                                                                                                  |
+  # |-------------|-----------|--------------|----------------------------------------------------------------------------------------------------------------------------------------|
+  # | no          | no        | use defaults | exhaust all possibilities: if it's missing (runApp does not specify); THEN if it's not in shinyApp appParts$options; THEN use defaults |
+  # | yes         | no        | use shinyApp | if it's missing (runApp does not specify); THEN if it's in shinyApp appParts$options; THEN use shinyApp                                |
+  # | no          | yes       | use runApp   | if it's not missing (runApp specifies), use those                                                                                      |
+  # | yes         | yes       | use runApp   | if it's not missing (runApp specifies), use those                                                                                      |
+  #
+  # I tried to make this as compact and intuitive as possible,
+  # given that there are four distinct possibilities to check
+  runOpts <- appParts$options
+  passedArgs <- as.list(match.call())[-1]
```

Comment:
I think the previous way, with the `missing` calls, had clearer intent. The `match.call()` and `environment()` assignment makes things kind of confusing. You could do something like this, where `f` is like `runApp`; `x` is like the app object; and, `arg1` is an argument like `port`:

```R
f <- function(x, arg1 = 1) {
  objValueOrDefault <- function(obj, name, default) {
    if (name %in% names(obj))
      obj[[name]]
    else
      default
  }
  
  if (missing(arg1)) arg1 <- objValueOrDefault(x, "arg1",  arg1)
  
  arg1
}
```

The drawback is that `arg1` appears 4 times on one line, and that the `if(missing())` part needs to be done for each argument. But it's easier to understand what it's doing.

It passes the following:

```R
library(testthat)
expect_identical(f(list()),              1)
expect_identical(f(list(), 2),           2)
expect_identical(f(list(arg1 = 3)),       3)
expect_identical(f(list(arg1 = 3), 2),    2)
expect_identical(f(list(arg1 = NULL)),    NULL)
expect_identical(f(list(arg1 = NULL), 2), 2)
```

The one tricky case is in the fifth test: the object has `arg1=NULL`, and that NULL gets used. I think that behavior makes the most sense.

If we didn't do things that way (if it returned `f`'s default value for `arg1`, which is 1), then `f` could be simplified:

```R

f <- function(x, arg1 = 1) {
  `%OR2%` <- function(x, y) {
    if (!is.null(x)) x
    else             y
  }
  
  if (missing(arg1)) arg1 <- x$arg1 %OR2% arg1
  
  arg1
}
```